feat(jsii-spec): Model parameter optionality

RomainMuller opened the pull request:
Method `Parameters`, `Properties` and method return types
now carry an `optional` flag that indicates whether they are
optional or required, and the `TypeReference#optional` field
was removed.

BREAKING CHANGE: JSII assemblies generated by older versions of the tool
will fail loading with this new version, and vice-versa. Re-compile your
projects in order to fix this.

Fixes #296
Fixes #414

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

RomainMuller commented:
Note that I decided against modeling *optional properties*, because it is essentially meaningless... A property can be considered optional as long as it's type is `nullable`.

rix0rrr commented on packages/jsii-kernel/lib/kernel.ts:
```diff
@@ -717,15 +717,15 @@ export class Kernel {
                     }
                 }
-            } else if (!param.type.optional && arg === undefined) {
+            } else if (!param.type.nullable && arg === undefined) {
```
This should be `param.optional` I think. This code is doing argument matching, not type matching.

rix0rrr commented on packages/jsii-kernel/lib/serialization.ts:
```diff
@@ -483,7 +483,7 @@ export function serializationType(typeRef: CompleteTypeReference, lookup: TypeLo
     // Propagate the top-level 'optional' field to each individual subtype
```
Miss comment.

rix0rrr commented on packages/jsii-pacmak/lib/targets/python.ts:
```diff
@@ -374,12 +360,12 @@ abstract class BaseMethod implements PythonBase {
                     const paramName = toPythonParameterName(prop.name);
                     const paramType = resolver.resolve(prop.type, { forwardReferences: false });
-                    const paramDefault = prop.type.optional ? "=None" : "";
+                    const paramDefault = prop.type.nullable ? "=None" : "";
```
This is not doing method arguments, right? A non-optional nullable type shouldn't have a default.

rix0rrr commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -451,10 +451,32 @@ export interface Parameter extends Documentable {
+export enum ParameterModifier {
+    /**
+     * Indicates that the parameter may be omitted. When a parameter has the
+     * `Optional` midufuer, the following must also be true:
```
fun typo :)

RomainMuller replied:
Oh there be many midufüren around here ...

rix0rrr commented on packages/jsii/lib/assembler.ts:
```diff
@@ -1175,6 +1162,28 @@ function _fingerprint(assembly: spec.Assembly): spec.Assembly {
+ * is otherwise difficult to assert whether a parameter with a default value is `Optional` or it it
+ * merely has a `nullable` `#type`.
+ */
+function _removeInvalidOptionalModifiers(parameters?: spec.Parameter[]) {
```
I see, is this the magic sauce that makes us backwards compatible.

Are we going to try looking at the declaration at all? Or too risky? (`name?: type` vs `name: type | undefined` ?)

rix0rrr commented on packages/jsii/lib/assembler.ts:
```diff
@@ -1175,6 +1162,28 @@ function _fingerprint(assembly: spec.Assembly): spec.Assembly {
+  function _cannotBeOptional(param: spec.Parameter, index: number, array: spec.Parameter[]): boolean {
+    return spec.isOptional(param)
+      // Checks whether there is one non-optional/variadic parameter after this one in the array
+      && array.find((p, i) => i > index && !spec.isOptional(p) && !spec.isVariadic(p)) != null;
```
Can be `array.some()`

RomainMuller replied:
Ugh I never find it because I keep thinking about the ruby name... `.any?`

rix0rrr commented:
It slightly concerns me that there are breaking changes in the assembly model. Do we know for a fact users will always use the new kernel to load the new assemblies?

Don't we have a `version` in the assembly metadata? Did you bump that, and does the kernel actually check the version it's expecting?

RomainMuller commented:
@rix0rrr new revision should have addressed all of your comments as well, and sets the Schema version of `jsii/0.9.0` (which I probably should flip to `0.10.0`). I wonder if I should get this to be expressive of a minimum version requirement, instead of looking like a fixed version... (can be a "for later" thing as well, but I guess if anyone has ideas...)

rix0rrr commented on packages/jsii-calc-base-of-base/test/assembly.jsii:
```diff
@@ -16,7 +16,7 @@
     "url": "https://github.com/awslabs/jsii.git"
   },
-  "schema": "jsii/1.0",
+  "schema": "jsii/0.9.0",
```
Waitaminit! I don't like how this is version number is going backwards, and we already have a `jsiiVersion` field that encodes the JSII version. Why can't we just discretely bump the schema versions 1 -> 2 -> 3 etc? Saves us having to do the whole "what's the next release version" dance, too.

RomainMuller replied:
I don't really care. I kinda felt it was nice to correlate that with the `jsii` version so you can, from an error (like "Whoosh the version is `jsii/0.9.0`, but I expected to find `jsii/0.42.12`), determine what you need to do. I however would only update that version number if/when the spec gets a breaking change (so it acts as a "minimum required version" of sorts.

rix0rrr commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -7,7 +7,7 @@ export interface Assembly extends Documentable {
      * The version of the spec schema
      */
-    schema: SchemaVersion.V1_0;
+    schema: SchemaVersion.LATEST;
```
I think if you're going to do this, it needs a version check in the jsii-kernel and jsii-reflect as well.

RomainMuller replied:
Both validate the schema, so the version check is in there.

rix0rrr commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -330,29 +330,38 @@ export enum PrimitiveType {
+/**
+ * A reference to a type (primitive, collection or fqn).
+ */
+export type TypeReference = NamedTypeReference | PrimitiveTypeReference | CollectionTypeReference | UnionTypeReference;
```
I support this!

rix0rrr commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -330,29 +330,38 @@ export enum PrimitiveType {
+ * Represents some instance of a type, for example as the return value of a method, a parameter, or a property.
  */
-export type TypeReference = TypeReferenceBase & (NamedTypeReference | PrimitiveTypeReference | CollectionTypeReference | UnionTypeReference);
+export interface TypeInstance<T extends TypeReference> {
```
I think I'm going to say that this interface ony makes sense for return types, and I think I object to its use in other places.

rix0rrr commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -390,7 +399,7 @@ export interface CollectionTypeReference extends TypeReferenceBase {
          * The type of an element (map keys are always strings).
          */
-        elementtype: TypeReference;
+        elementtype: TypeInstance<TypeReference>;
```
It does not make sense for either array or map elements to be of type `Promise<T>`, or optional.

RomainMuller replied:
So an array may never contain `null`? Or it'll contain "surprise" `null`s?

rix0rrr replied:
But that is something else though! Feels like now we're expressing nullability, not optionality.

rix0rrr commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -409,7 +418,7 @@ export interface UnionTypeReference extends TypeReferenceBase {
          * @minItems 2
          */
-        types: TypeReference[];
+        types: Array<TypeInstance<TypeReference>>;
```
It does not make sense for a union type to be something like `number | string?` or `number | Promise<string>`.

RomainMuller replied:
And yet, this is valid TypeScript, and thus could be authored in a TypeScript signature that JSII would totally parse out in that way

rix0rrr replied:
Nobody says we can't reject that in jsii. Will the kernel handle it correctly?

rix0rrr commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -484,13 +493,17 @@ export interface Parameter extends Documentable {
+     * The specification for the value of this parameter (includes it's declared type and whether it is optional).
      */
-    type: TypeReference;
+    value: TypeInstance<TypeReference>;
```
A parameter can never be a `Promise<T>`. It *can* be optional, however.

RomainMuller replied:
```ts
// In fact, totally possible:
async function impossible(arg: Promise<Foo>): Promise<Bar>;
```

rix0rrr replied:
Sure. What I meant to say is that the kernel does not have adequate provisions to handle this, so we shouldn't allow it either.

rix0rrr commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -508,7 +521,7 @@ export interface Method extends Documentable, Overridable, SourceLocatable {
      * The return type of the method (undefined if void or initializer)
      */
-    returns?: TypeReference;
+    returns?: TypeInstance<TypeReference>;
```
This is the only place where `TypeInstance` makes sense to me.

Although I think `promise?: boolean` makes more sense as an attribute of `Method` to me (`async?: boolean`) than as an attribute of the return type.

rix0rrr commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -440,9 +449,9 @@ export interface Property extends Documentable, Overridable, SourceLocatable {
+     * The specification for the property's value (including it's declared type and whether it is optional).
      */
-    type: TypeReference;
+    value: TypeInstance<TypeReference>;
```
Properties cannot be `promise`. They *can* be *optional* however.

RomainMuller replied:
I can totes declare a `public readonly iSwearItWillHaveAValueAtSomePoint: Promise<string>`. If that's an API I *like*? No. But it can be made (and JSII will happily generate a `promise: true` for that type reference.

rix0rrr replied:
See my previous comment, I'm not sure the kernel will handle that situation properly.

rix0rrr commented:
I think I'd prefer a modeling like this:

```ts
interface OptionalValue {
    optional?: boolean;
    type: TypeReference;
}

interface Method {
    async?: boolean;
    returns?: OptionalValue;
    // ...
}

interface Property extends OptionalValue {
    // ...
}

interface Collection {
    element: TypeReference;
    // ...
}

```

eladb commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -324,40 +412,53 @@ export enum PrimitiveType {
+ * The standard representation of the `any` type (includes optionality marker).
  */
-export interface TypeReferenceBase {
+export const CANONICAL_ANY: Readonly<PrimitiveTypeReference> = { optional: true, primitive: PrimitiveType.Any };
```
Why does `any` imply `optional`? It implies nullable, but that's a different thing.

RomainMuller replied:
I guess I think of optional as allowing "no value" through, and any also allows "no value" though. But I guess I see your point still...

eladb commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -324,40 +412,53 @@ export enum PrimitiveType {
+/**
+ * Qualifiers for a type reference.
+ */
+interface QualifiedTypeReference {
```
Maybe `TypeReferenceQualifiers`, because this is not a type reference.

eladb commented on packages/jsii-pacmak/test/expected.jsii-calc/python/src/jsii_calc/__init__.py:
```diff
@@ -58,7 +58,7 @@ def __init__(self) -> None:
 
     @jsii.member(jsii_name="anyIn")
-    def any_in(self, inp: typing.Any) -> None:
+    def any_in(self, inp: typing.Any=None) -> None:
```
That's wrong... any does not imply optionality

RomainMuller commented on packages/jsii/lib/assembler.ts:
```diff
@@ -555,17 +557,17 @@ export class Assembler implements Emitter {
       // tslint:disable-next-line:no-bitwise
       if ((ts.getCombinedModifierFlags(ctorDeclaration) & ts.ModifierFlags.Private) === 0) {
-        jsiiType.initializer = { initializer: true };
+        jsiiType.initializer = {};
```
You have an initializer, not a method. Methods have a name (guaranteed). Initializers don't (guaranteed).

RomainMuller commented on packages/jsii/lib/assembler.ts:
```diff
@@ -204,22 +203,26 @@ export class Assembler implements Emitter {
    */
-  private _dereference(ref: spec.NamedTypeReference, referencingNode: ts.Node | null): spec.Type | undefined {
-    const [assm, ] = ref.fqn.split('.');
+  private _dereference(ref: string | spec.NamedTypeReference, referencingNode: ts.Node | null): spec.Type | undefined {
```
Oh boy. That alias even exists! I can just use it :)

RomainMuller commented on packages/jsii-calc-base-of-base/test/assembly.jsii:
```diff
@@ -35,15 +35,14 @@
-      "initializer": {
-        "initializer": true
-      },
+      "initializer": {},
```
Empty object here is the default initializer (no parameters, public). I thought of also modeling that as "true" or the literal string "default" instead... but the empty object actually describes what I want, based on its default values... so I figured I wouldn't overthink it.

eladb replied:
I understand where this is coming from, but still don't like this as a general modeling principle. Can we do better?

RomainMuller commented:
@rix0rrr well then you seem in vivid disagreement with @eladb and I am certainly not re-doing this whole thing a fourth time unless we are 100% certain the 4th is the right one and it has significant benefits over this... this is a full day of painful delving into C# failing tests each time...

eladb commented:
@RomainMuller what disagreement?

RomainMuller commented:
@eladb isn't what @rix0rrr "would prefer" a minor variation over the previous iteration , which I would then literally have aster a whole day redoing all over again?

RomainMuller commented:
I want to be able to express:
- optional result
- optional parameter (as in can be undefined)
- optional value in collection (list/map)

eladb commented:
Is it possible to decouple these discussions? We are in alignment on optionals, so let’s wrap up this PR, and follow up with a discussion and a separate PR on nullables. I don’t see a good reason to couple those together, right?

RomainMuller commented:
Additional change discussions will happen in #440

rix0rrr commented:
> Is it possible to decouple these discussions?

I don't think so. We currently have one flag which we interpret as "optional" in a value context (property, argument) and as "nullable" in a type context (return values, `Array<Nullable<string>>`).

I would suggest we either make a decision on what properties we want from our type system and encode those, or we leave the current situation which is occasionally confusing but serviceable enough.

There is only one thing we cannot model today, which is a required, nullable parameter.

eladb commented on packages/jsii-spec/lib/spec.ts:
```diff
@@ -210,34 +239,46 @@ export interface Docs {
     /**
      * Whether the API item is beta/experimental quality
+     *
+     * @default Stability.Experimental
```
inherit from parent (and if no parent - stable)
